external-crates/move/crates/move-package/src/lock_file/schema.rs

```diff
@@ -110,6 +110,25 @@ impl Packages {
     }
 }
 
+impl ToolchainVersioning {
+    /// Read toolchain versioning info from the lock file. Returns successfully with None if
+    /// parsing the lock file succeeds but an entry for `[toolchain-version]` does not exist.
+    pub fn read(lock: &mut impl Read) -> Result<Option<ToolchainVersioning>> {
+        let mut buf = String::new();
+        lock.read_to_string(&mut buf).context("Reading lock file")?;
+        let toml: toml::Value = toml::from_str(&buf)?;
+        let move_ = toml.get("move").ok_or_else(|| anyhow!("No [move] found"))?;
+        match move_.get("toolchain-versioning") {
+            None => Ok(None), // parsing succeeds, no entry for `toolchain-versioning`
+            Some(v) => v
+                .clone()
+                .try_into()
+                .map(Some)
+                .map_err(|e| anyhow!("Error deserializing: {e}")),
+        }
+    }
```

Comment:
nit: Slight discrepancy between the comment, which mentions `[toolchain-version]` and the actual field name which is `[toolchain-versioning]` -- we should reconcile that.

FWIW I prefer `[toolchain-version]`, I wasn't going to mention it but now that I see you are conflicted about it, I thought I'd mention 😅.

On the implementation -- this pattern is why `Schema` has a generic type.  You should be able to do something like the following, to avoid hand rolling the code to explore the TOML tree:

```rust
impl ToolchainVersion {
    pub fn read(lock: &mut impl Read) -> Result<Option<ToolchainVersion>> {
        let contents = {
            let mut buf = String::new();
            lock.read_to_string(&mut buf).context("Reading lock file")?;
            buf
        };
        
        struct TV(#[serde(rename = "toolchain-version")] Option<ToolchainVersion>);
        let Schema { move_: toolchain_version } = toml::de::from_str::<Schema<TV>>(&contents)
            .context("Deserializing toolchain version")?;

         Ok(toolchain_version.0)
    }
}
```

But I guess you've seen the pattern for `Packages ` already, so you're going to burst my bubble and tell me this doesn't work...

Reply:
> But I guess you've seen the pattern for `Packages` already, so you're going to burst my bubble and tell me this doesn't work...

I did try and I couldn't figure out the incantation :-) With your help I got it, needed to also tweak the type to be a struct with a field named `toolchain-versioning` and `derive` on it:

```rust
        #[derive(Deserialize)]
        struct TV {
            #[serde(rename = "toolchain-versioning")]
            toolchain_versioning: Option<ToolchainVersioning>,
        }
        let Schema { move_: value } = toml::de::from_str::<Schema<TV>>(&contents)
            .context("Deserializing toolchain version")?;
```

Purely as a remark (and not really in defense of anything), I experience this kind of thing to be a bit "too" magical, and the previous hand roll is much more obvious to me what's going on. And, evidence that I could find a way to do what I was trying to do :P

With the ser/deser magic we're dealing with inline type defs, macros, destructuring with generic types--in the previous, simply casing out on the structure. I understand the rationale behind using the "magic" version, the safety, maybe terseness, I really do, but comparatively it's... a lot. I experienced the same kind of internal conflict with how axum middleware and responses work--far too much magic to figure out how to use the abstraction. Anyway! Just remarking on my journey of getting better at magic Rust. Thanks for the help!

Reply:
Indeed -- these libraries (`serde`, `axum`, etc) are really whole languages unto themselves, so you trick yourself over time into thinking the complexity is not there because it becomes innate, but I don't disagree with you!